src/librustc_typeck/check/_match.rs

```diff
@@ -212,7 +216,19 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
                             let mt = ty::TypeAndMut { ty: inner_ty, mutbl: mutbl };
                             let region = self.next_region_var(infer::PatternRegion(pat.span));
                             let rptr_ty = tcx.mk_ref(region, mt);
-                            self.demand_eqtype(pat.span, expected, rptr_ty);
+                            let err = self.demand_eqtype_diag(pat.span, expected, rptr_ty);
+                            if let Some(mut err) = err {
+                                if is_arg {
+                                    if let Ok(snippet) = self.sess().codemap()
+                                        .span_to_snippet(pat.span)
+                                    {
+                                        err.help(&format!("did you mean `{}: &{}`?",
```

Comment:
One concern -- what happens in more extreme cases? I think we should check that the inner pattern (represented by `inner`, here) is a binding pattern, so as not to trigger for something like `fn foo(&&bar: &u32)` or `fn foo(&[bar]: &u32)`. Suggesting `&bar: &&u32` and `[bar]: &&u32` respectively doesn't seem all that helpful (that is what we will do, right?)